lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,149 @@
+          o.banner = BANNER
+          o.on('--config', 'Path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        _, certnames, invalid = Utils.parse_without_raising(parser, args)
+
+        exit_code = nil
+        unless invalid.empty?
+          Utils.print_invalid_args(@logger, invalid, parser.help)
+          exit_code = 1
+        end
+
+        case certnames.length
+        when 0
+          @logger.err 'A certname to revoke is required'
+          @logger.err parser.help
+          exit_code = 1
+        when 1
+          results[:certname] = certnames.first
+        else
+          @logger.err 'Only one certname is allowed at a time'
+          @logger.err 'Received certnames:'
+          certnames.each {|name| @logger.err "    #{name}" }
+          exit_code = 1
+        end
+
+        return results, exit_code
+      end
+
+      # take the parsed arg hash from #parse and perform this action
+      # print information to @logger and return a valid exit code
+      def run(args)
+        certname = args[:certname]
+        config = args[:config]
+
+        errors = validate_file_paths(config)
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|error| @logger.err "    #{error}" }
+          exit 1
+        end
+
+        puppet = PuppetConfig.parse(config)
+
+        revoke(certname, puppet.settings)
+      end
+
+      def make_store(cacert, cert, crl)
+        store = OpenSSL::X509::Store.new
+        store.add_file(cacert)
+        store.add_file(cert)
+        store.add_file(crl)
+
+        store
+      end
+
+      def make_uri(host, port, certname)
+        URI('https://' + host + ':' + port +
+            '/puppet-ca/v1/certificate_status/' +
+            certname +
+            '?environment=production')
+      end
+
+      def revoke(certname, settings)
+        body = JSON.dump({ desired_state: 'revoked' })
+
+        headers = {
+          'User-Agent' => 'PuppetserverCaCli',
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json'
+        }
+
+        store = make_store(settings[:localcacert],
+                           settings[:hostcert],
+                           settings[:hostcrl])
+
+        uri = make_uri(settings[:ca_server],
+                       settings[:ca_port],
+                       certname)
+
+        result = Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store) do |connection|
+          request = Net::HTTP::Put.new(uri, headers)
+          request.body = body
+          connection.request(request)
+        end
+
+        if result.is_a? Net::HTTPSuccess
+          @logger.inform "Revoked certificate for #{certname}"
+          return 0
+        elsif result.is_a? Net::HTTPNotFound
```

Comment:
turns out cert doesn't do much for errors, afaict

Reply:
One thing it does is return the serial number rather than the certname, which the API doesn't do. I don't know if that's super valuable or not (it becomes important when `all_duplicate_certnames` is `true`).

Reply:
Okay, it looks like the current tool (at least according to its documentation) allows for specifying a cert to revoke either by its certname or its serial number (in hex format):
```
be puppet cert --help
..snip..
* revoke:
  Revoke the certificate of a client. The certificate can be specified either
  by its serial number (given as a hexadecimal number prefixed by '0x') or by its
  hostname. The certificate is revoked by adding it to the Certificate Revocation
  List given by the 'cacrl' configuration option. Note that the puppet master
  needs to be restarted after revoking certificates.
```

However the API [doesn't let you do this](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/puppetserver/certificate_authority.clj#L1260-L1263), and doesn't return the serial of the revoked cert.

I think we might want to add this functionality later - I assume its priority would depend on the usage of `allow_duplicate_certnames` in the field? Should we make a ticket for that @Magisus ?

Reply:
Yeah I'll file one.